src/services/api/api.contoller.test.ts

```diff
@@ -0,0 +1,189 @@
+import { Test } from '@nestjs/testing';
+import { MongooseModule } from '@nestjs/mongoose';
+import { ConfigModule } from '@nestjs/config';
+import { INestApplication, ValidationPipe, Module, CanActivate, ExecutionContext, CallHandler, Injectable } from '@nestjs/common';
+import request from 'supertest';
+import { MongoMemoryServer } from 'mongodb-memory-server';
+import { AuthGuard } from '@nestjs/passport';
+import { ApiController } from './api.controller';
+import { ApiModule } from './api.module';
+import { AuthService } from '../auth/auth.service';
+import { HiveAccountRepository } from '../../repositories/hive-account/hive-account.repository';
+import { UserRepository } from '../../repositories/user/user.repository';
+import { HiveRepository } from '../../repositories/hive/hive.repository';
+import { LinkedAccountRepository } from '../../repositories/linked-accounts/linked-account.repository';
+import { EmailService } from '../email/email.service';
+import { HiveAccountModule } from '../../repositories/hive-account/hive-account.module';
+import { UserModule } from '../../repositories/user/user.module';
+import { LinkedAccountModule } from '../../repositories/linked-accounts/linked-account.module';
+import { JwtModule } from '@nestjs/jwt';
+import { AuthModule } from '../auth/auth.module';
+import { MockAuthGuard, MockUserDetailsInterceptor, UserDetailsInterceptor } from './utils';
+import { HiveModule } from '../../repositories/hive/hive.module';
+import { EmailModule } from '../email/email.module';
+
+describe('ApiController', () => {
+  let app: INestApplication;
+  let mongod;
+  let authService: AuthService;
+  let hiveAccountRepository: HiveAccountRepository;
+  let userRepository: UserRepository;
+  let hiveRepository: HiveRepository;
+  let linkedAccountsRepository: LinkedAccountRepository;
+  let emailService: EmailService;
+
+  beforeAll(async () => {
```

Comment:
Unless you are checking state side effects between tests, I'd recommend constructing a clean mock instance for each test.

```suggestion
  beforeEach(async () => {
```

Reply:
Thats not the recommended approach according to the docs for e2e tests: https://docs.nestjs.com/fundamentals/testing
I think resetting just the DB would probably have the same effect? Does seem overkill to take the entire API down and putting it back up several times

Reply:
The docs don't actually mention 1 method vs the other. It just has an example that happens to use `beforeAll`.

Unless you are 100% sure that all the Nest.JS classes have 0 internal side-effect (i.e. `this.something = 'something else'`) when you call their corresponding REST endpoints, I would not recommend doing e2e testing like this. This method, while it may take less time to execute, can lead to faulty tests.

Reply:
Here's a few reasons why its preferred for e2es vs unit tests and probably why nest used it for other e2e examples, the issues you raise are more likely than not, fixable by resetting the in-memory DB:

### Performance

- Startup Time: Initializing the NestJS application can be time-consuming, especially if it involves setting up a database connection, loading modules, and performing other initialization tasks. Doing this setup once in beforeAll reduces the overhead compared to setting up before each test.

- Resource Intensive: For larger applications, re-initializing the application for each test can be resource-intensive and slow down the test suite significantly.
Typical E2E Test Scope:

- Broad Scope: E2E tests are typically designed to test the entire application flow from start to finish, rather than isolated units of functionality. Therefore, they often require a fully initialized application.

- Full Application Context: Since E2E tests aim to simulate real user interactions with the full application, having the application context available for all tests makes sense.

### Consistency:

- Shared Application State: In some E2E test scenarios, it is acceptable or even desirable to have a shared application state, especially when testing workflows that span multiple requests or sessions.

- Stable Environment: Keeping the application running throughout the test suite can lead to a more stable testing environment, avoiding potential issues with repeatedly starting and stopping the application.

Reply:
I understand what I'm suggesting can have major performance impact on test execution speed, but I'd argue that it does not matter. The only time test suite speed matter is if it blocks the development workflow for the project via something like CI/CD.

I think the term "E2E test" is being conflated with "integration test". For sake of argument, let's define them as follows:

"E2E test": automated test that simulates a complete user flow. For example, user logins, likes a post, then logs out.

"integration test": automated test that tests a specific endpoint or API in isolation simulating how an external service might use the API.

Using the above definitions, it seems like you have written tests more similar to integration tests rather than E2E tests. For these tests, I think it makes more sense to have a clean state at the beginning of each test to ensure consistent results for each test.

If you'd still like to do E2E tests, I would recommend abstracting the API requests into realistic user flows instead and doing multiple rounds of tests on those flows to ensure consistency without resetting the test environment every time, but this would be separate from what you are doing here.